### ISSUE TITLE
Something is wrong with optimizer

### ISSUE THREAD

**Issue by dmuraco3** [dmuraco3/rust_grad#1]
https://github.com/dmuraco3/rust_grad/blob/7035dce3c84ce85c86b5b82e654ab29a5315e42d/src/nn/optim/mod.rs#L1-L148

Loss barely changes after optimizer.step()

I'm guessing there is a problem with the tensor data held by the optimizer.

The optimizer should take in references to the weights, then clone the tensors to own them. When cloning the tensors, Arc::clone should be called on the underlying Arc<RwLock<Storage::Vec>>, thereby preserving the mutable data.

I feel like either my optimizer code isn't correct or the underlying weight tensors' data is being cloned to new memory.

Will investigate soon.

**Comment by dmuraco3** [comment]
https://github.com/dmuraco3/rust_grad/blob/7035dce3c84ce85c86b5b82e654ab29a5315e42d/src/nn/optim/mod.rs#L1-L148
